viabel/approximations.py

```diff
@@ -372,3 +374,108 @@ def supports_pth_moment(self, p):
     def df(self):
         """Degrees of freedom."""
         return self._df
+
+
+class NeuralNet(ApproximationFamily):
+    def __init__(self, layers_shapes, nonlinearity = np.tanh):
+        self._pattern = PatternDict(free_default = True)
+        self._layers = len(layers_shapes)
+        self._nonlinearity = nonlinearity
+        self.input_dim = layers_shapes[1][1]
+        for l in range(len(layers_shapes)):
+            self._pattern[str(l)] = NumericArrayPattern(shape = layers_shapes[l])
+
+    def forward(self, var_param, x):
+        for l in range(self._layers):
+            W = var_param[str(l)]
+            x = np.dot(x, W)
+            x = np.where(x < 0, 1e-2 * x, x)
+        return self._nonlinearity(x)
+
+    def sample(self, var_param, n_samples):
+        x = npr.multivariate_normal(mean = [0] * self.input_dim,
+                                    cov = np.identity(self.input_dim),
+                                    size = n_samples)
+        z = self.forward(x)
+        return z
+
+    def log_density(self, var_param, x):
```

Comment:
please add implementation

Reply:
I'm implementing this but got into a bit of a wall. Because the BNN is modelling an implicit distribution I think we can just evaluate the prior over the variational parameters? There's some work regarding functional BNNs which could be more flexible as they don't impose a prior on the weights themselves but on the function space https://openreview.net/pdf?id=rkxacs0qY7.

Reply:
I don't know what you mean by prior. If f_lambda is the neural net, then the r.v. is Z = f_lambda(Z_0) for Z_0 ~ N(0, I), so you can compute the density of Z as long as the Jacobian of f_lambda is available, which it is for NFs.